File: baekjoon/data_structure/4949.js
function isBalanced(line) {
  const chars = line.split('');
  const counters = [];

  for (let i = 0; i < chars.length; i++) {
    const char = chars[i];

    switch (char) {
      case '.':
        if (counters.length > 0) return false;
        else return true;
      case '(':
        counters.push(')');
        break;
      case '[':
        counters.push(']');
        break;
      case ')':
        if (counters.pop() !== char) return false;
        break;
      case ']':
        if (counters.pop() !== char) return false;
        break;
      default:
        break;
    }
  }

  return true;
}

function solution(input) {
  let ans = [];

  input.forEach(line => {
    isBalanced(line) ? ans.push('yes') : ans.push('no');
  });

  return ans.join('\n');
}

function main() {
  const fs = require('fs');
  const input = fs
    .readFileSync('./baekjoon/data_structure/4949.txt')
    .toString()
    .trim()
    .split('\n')
    .filter(v => v !== '.');

  console.log(solution(input));
}

if (require.main === module) main();

File: baekjoon/data_structure/4949.py
from sys import stdin
from typing import List


def solution(string: str):
    stack = []

    for s in string:
        if s == "(":
            stack.append(s)
        elif s == ")":
            if validate("(", stack):
                stack.pop()
                continue
            else:
                print("no")
                return

        elif s == "[":
            stack.append(s)
        elif s == "]":
            if validate("[", stack):
                stack.pop()
                continue
            else:
                print("no")
                return

    if len(stack) <= 0:
        print("yes")
    else:
        print("no")


def validate(target_bracket: chr, stack: List[chr]) -> bool:
    if len(stack) <= 0:
        return False

    if stack[-1] != target_bracket:
        return False

    return True


if __name__ == "__main__":
    for line in stdin:
        if line == ".\n" or line == ".":
            break
        string = []
        for s in line:
            if s == "(" or s == ")" or s == "[" or s == "]":
                string.append(s)
        solution(string)
